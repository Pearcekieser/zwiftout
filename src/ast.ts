import { IntervalType } from "./parser/tokenizer";
import { Duration } from "./Duration";
import { Intensity } from "./Intensity";

export type Workout = {
  name: string;
  author: string;
  description: string;
  tags: string[];
  intervals: Interval[];
};

export type Interval = {
  type: IntervalType;
  duration: Duration;
  intensity: Intensity;
  cadence?: number;
  comments: Comment[];
};

export type Comment = {
  offset: Duration;
  text: string;
};
