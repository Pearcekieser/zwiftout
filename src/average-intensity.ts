import { pipe, sum } from "ramda";
import { Interval } from "./ast";
import { intervalsToIntensities } from "./intervalsToIntensities";

const average = (arr: number[]) => sum(arr) / arr.length;

export const averageIntensity = (intervals: Interval[]): number => {
  return pipe(intervalsToIntensities, average)(intervals);
};
