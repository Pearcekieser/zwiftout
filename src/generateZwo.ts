import * as xml from "xml";
import { Interval, Workout, Comment } from "./ast";
import { detectRepeats, RepeatedInterval } from "./detectRepeats";

// Zwift Workout XML generator

const generateTextEvents = (comments: Comment[]): xml.XmlObject[] => {
  return comments.map(({ offset, text }) => ({
    textevent: [{ _attr: { timeoffset: offset.seconds, message: text } }],
  }));
};

const generateRangeInterval = (
  tagName: "Warmup" | "Cooldown" | "Ramp",
  { duration, intensity, cadence, comments }: Interval,
): xml.XmlObject => {
  return {
    [tagName]: [
      {
        _attr: {
          Duration: duration.seconds,
          PowerLow: intensity.start,
          PowerHigh: intensity.end,
          ...(cadence ? { Cadence: cadence } : {}),
        },
      },
      ...generateTextEvents(comments),
    ],
  };
};

const generateSteadyStateInterval = ({ duration, intensity, cadence, comments }: Interval): xml.XmlObject => {
  return {
    SteadyState: [
      {
        _attr: {
          Duration: duration.seconds,
          Power: intensity.value,
          ...(cadence ? { Cadence: cadence } : {}),
        },
      },
      ...generateTextEvents(comments),
    ],
  };
};

const generateFreeRideInterval = ({ duration, comments }: Interval): xml.XmlObject => {
  return {
    FreeRide: [
      {
        _attr: {
          Duration: duration.seconds,
        },
      },
      ...generateTextEvents(comments),
    ],
  };
};

const generateRepeatInterval = (repInterval: RepeatedInterval): xml.XmlObject => {
  const [on, off] = repInterval.intervals;
  return {
    IntervalsT: [
      {
        _attr: {
          Repeat: repInterval.times,

          OnDuration: on.duration.seconds,
          OnPower: on.intensity.start,
          ...(on.cadence ? { Cadence: on.cadence } : {}),

          OffDuration: off.duration.seconds,
          OffPower: off.intensity.end,
          ...(off.cadence ? { CadenceResting: off.cadence } : {}),
        },
      },
      ...generateTextEvents(repInterval.comments),
    ],
  };
};

const generateInterval = (
  interval: Interval | RepeatedInterval,
  index: number,
  allIntervals: (Interval | RepeatedInterval)[],
): xml.XmlObject => {
  if (interval.type === "repeat") {
    return generateRepeatInterval(interval);
  }

  const { intensity } = interval;
  if (index === 0 && intensity.start < intensity.end) {
    return generateRangeInterval("Warmup", interval);
  } else if (index === allIntervals.length - 1 && intensity.start > intensity.end) {
    return generateRangeInterval("Cooldown", interval);
  } else if (intensity.start !== intensity.end) {
    return generateRangeInterval("Ramp", interval);
  } else if (intensity.zone === "free") {
    return generateFreeRideInterval(interval);
  } else {
    return generateSteadyStateInterval(interval);
  }
};

const generateTag = (name: string): xml.XmlObject => {
  return {
    tag: [{ _attr: { name } }],
  };
};

export const generateZwo = ({ name, author, description, tags, intervals }: Workout): string => {
  return xml(
    {
      workout_file: [
        { name: name },
        { author: author },
        { description: description },
        { tags: tags.map(generateTag) },
        { sportType: "bike" },
        { workout: detectRepeats(intervals).map(generateInterval) },
      ],
    },
    { indent: "  " },
  );
};
