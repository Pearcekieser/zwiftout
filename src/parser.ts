import { Interval, Workout } from "./ast";
import { LabelTokenValue, Token } from "./tokenizer";

type Header = {
  name?: string;
  author?: string;
  description?: string;
};

const extractText = (tokens: Token[]): [string, Token[]] => {
  let text;
  while (tokens[0] && tokens[0].type === "text") {
    if (text === undefined) {
      text = tokens[0].value;
    } else {
      text += "\n" + tokens[0].value;
    }
    tokens.shift();
  }
  return [text || "", tokens];
};

const parseHeader = (tokens: Token[]): [Header, Token[]] => {
  const header: Header = {};

  while (tokens[0]) {
    const token = tokens[0];
    if (token.type === "label" && token.value === LabelTokenValue.Name) {
      tokens.shift();
      const [name, rest] = extractText(tokens);
      header.name = name;
      tokens = rest;
    } else if (
      token.type === "label" &&
      token.value === LabelTokenValue.Author
    ) {
      tokens.shift();
      const [author, rest] = extractText(tokens);
      header.author = author;
      tokens = rest;
    } else if (
      token.type === "label" &&
      token.value === LabelTokenValue.Description
    ) {
      tokens.shift();
      const [description, rest] = extractText(tokens);
      header.description = description;
      tokens = rest;
    } else {
      // End of header
      break;
    }
  }

  return [header, tokens];
};

const parseIntervals = (tokens: Token[]): Interval[] => {
  return [];
};

export const parse = (tokens: Token[]): Workout => {
  const [header, intervalTokens] = parseHeader(tokens);
  return {
    name: header.name || "",
    author: header.author || "",
    description: header.description || "",
    intervals: parseIntervals(intervalTokens),
  };
};
