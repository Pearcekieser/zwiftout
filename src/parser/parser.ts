import { Interval, Workout, Comment } from "../ast";
import { Duration } from "../Duration";
import { ConstantIntensity, FreeIntensity, RangeIntensity } from "../Intensity";
import { ParseError } from "./ParseError";
import { IntervalType, SourceLocation, Token } from "./tokenizer";

type Header = Partial<Omit<Workout, "intervals">>;

const tokenToString = (token: Token | undefined): string => {
  return token ? `[${token.type} ${token.value}]` : "EOF";
};

const extractText = (tokens: Token[]): [string, Token[]] => {
  let text;
  while (tokens[0] && tokens[0].type === "text") {
    if (text === undefined) {
      text = tokens[0].value;
    } else {
      text += "\n" + tokens[0].value;
    }
    tokens.shift();
  }
  return [text ? text.trim() : "", tokens];
};

const parseHeader = (tokens: Token[]): [Header, Token[]] => {
  const header: Header = {};

  while (tokens[0]) {
    const token = tokens[0];
    if (token.type === "header" && token.value === "Name") {
      tokens.shift();
      const [name, rest] = extractText(tokens);
      header.name = name;
      tokens = rest;
    } else if (token.type === "header" && token.value === "Author") {
      tokens.shift();
      const [author, rest] = extractText(tokens);
      header.author = author;
      tokens = rest;
    } else if (token.type === "header" && token.value === "Description") {
      tokens.shift();
      const [description, rest] = extractText(tokens);
      header.description = description;
      tokens = rest;
    } else if (token.type === "header" && token.value === "Tags") {
      tokens.shift();
      const [tags, rest] = extractText(tokens);
      header.tags = tags.split(/\s*,\s*/);
      tokens = rest;
    } else {
      // End of header
      break;
    }
  }

  return [header, tokens];
};

const parseIntervalComments = (tokens: Token[], intervalDuration: Duration): [Comment[], Token[]] => {
  const comments: Comment[] = [];
  while (tokens[0]) {
    const [start, offset, text, ...rest] = tokens;
    if (start.type === "comment-start") {
      if (!offset || offset.type !== "duration") {
        throw new ParseError(
          `Expected [comment offset] instead got ${tokenToString(offset)}`,
          offset?.loc || start.loc,
        );
      }
      if (!text || text.type !== "text") {
        throw new ParseError(`Expected [comment text] instead got ${tokenToString(text)}`, text?.loc || offset.loc);
      }
      comments.push({
        // when offset is negative, recalculate it based on interval length
        offset: new Duration(offset.value >= 0 ? offset.value : intervalDuration.seconds + offset.value),
        text: text.value,
        loc: offset.loc,
      });
      tokens = rest;
    } else {
      break;
    }
  }
  return [comments, tokens];
};

const parseIntervalParams = (type: IntervalType, tokens: Token[], loc: SourceLocation): [Interval, Token[]] => {
  let duration;
  let cadence;
  let intensity;

  while (tokens[0]) {
    const token = tokens[0];
    if (token.type === "duration") {
      duration = new Duration(token.value);
      tokens.shift();
    } else if (token.type === "cadence") {
      cadence = token.value;
      tokens.shift();
    } else if (token.type === "intensity") {
      intensity = new ConstantIntensity(token.value);
      tokens.shift();
    } else if (token.type === "intensity-range") {
      intensity = new RangeIntensity(token.value[0], token.value[1]);
      tokens.shift();
    } else {
      break;
    }
  }

  if (!duration) {
    throw new ParseError("Duration not specified", loc);
  }
  if (!intensity) {
    intensity = new FreeIntensity();
  }

  const [comments, rest] = parseIntervalComments(tokens, duration);

  return [{ type, duration, intensity, cadence, comments }, rest];
};

const parseIntervals = (tokens: Token[]): Interval[] => {
  const intervals: Interval[] = [];

  while (tokens[0]) {
    const token = tokens.shift() as Token;
    if (token.type === "interval") {
      const [interval, rest] = parseIntervalParams(token.value, tokens, token.loc);
      intervals.push(interval);
      tokens = rest;
    } else {
      throw new ParseError(`Unexpected token ${tokenToString(token)}`, token.loc);
    }
  }

  return intervals;
};

export const parseTokens = (tokens: Token[]): Workout => {
  const [header, intervalTokens] = parseHeader(tokens);

  return {
    name: header.name || "Untitled",
    author: header.author || "",
    description: header.description || "",
    tags: header.tags || [],
    intervals: parseIntervals(intervalTokens),
  };
};
