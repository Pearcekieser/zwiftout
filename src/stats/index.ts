import { Workout } from "../ast";
import { averageIntensity } from "./averageIntensity";
import { normalizedIntensity } from "./normalizedIntensity";
import { totalDuration } from "./totalDuration";
import { tss2 } from "./tss2";
import { zoneDistribution } from "./zoneDistribution";

// Generates statistics
export const stats = ({ intervals }: Workout): string => {
  const duration = totalDuration(intervals);
  const avgIntensity = averageIntensity(intervals);
  const normIntensity = normalizedIntensity(intervals);
  const zones = zoneDistribution(intervals);

  return `
Total duration: ${(duration.seconds / 60).toFixed()} minutes

Average intensity: ${(avgIntensity.value * 100).toFixed()}%
Normalized intensity: ${(normIntensity.value * 100).toFixed()}%

TSS: ${tss2(duration, normIntensity).toFixed()}

Zone Distribution:
${zones.map(({ name, duration }) => `${(duration.seconds / 60).toFixed().padStart(3)} min - ${name}`).join("\n")}
`;
};
