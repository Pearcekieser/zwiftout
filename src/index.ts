import * as fs from "fs";
import { parseFile } from "./parser";

const filename = process.argv[2];

console.log(`Parsing: ${filename}`);

const file = fs.readFileSync(filename, "utf8");

parseFile(file).forEach((rule) => {
  console.log(rule.type);
  rule.params.forEach((p) => {
    console.log(`  ${p.type}: ${p.value}`);
  });
});
