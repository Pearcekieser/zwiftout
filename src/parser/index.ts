import { Workout } from "../ast";
import { parseTokens } from "./parser";
import { tokenize } from "./tokenizer";

export const parse = (source: string): Workout => parseTokens(tokenize(source));
