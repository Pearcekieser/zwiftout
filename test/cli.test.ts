import * as fs from "fs";
import { generateZwo } from "../src/generateZwo";
import { parse } from "../src/parser";
import { formatStats, stats } from "../src/stats";

const createStats = (filename: string) => formatStats(stats(parse(fs.readFileSync(filename, "utf8"))));
const createZwo = (filename: string) => generateZwo(parse(fs.readFileSync(filename, "utf8")));

const filenames = [
  "examples/comments.txt",
  "examples/darth-vader.txt",
  "examples/ftp-test.txt",
  "examples/halvfems.txt",
  "examples/threshold-pushing.txt",
  "examples/ramps.txt",
];

describe("Generate ZWO", () => {
  filenames.forEach((filename) => {
    it(filename, () => {
      expect(createZwo(filename)).toMatchSnapshot();
    });
  });
});

describe("Generate stats", () => {
  filenames.forEach((filename) => {
    it(filename, () => {
      expect(createStats(filename)).toMatchSnapshot();
    });
  });
});
