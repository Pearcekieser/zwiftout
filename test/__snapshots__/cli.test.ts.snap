// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate ZWO examples/comments.txt 1`] = `
"<workout_file>
  <name>Workout with comments</name>
  <author></author>
  <description></description>
  <sportType>bike</sportType>
  <workout>
    <Warmup Duration=\\"600\\" PowerLow=\\"0.25\\" PowerHigh=\\"0.75\\">
      <textevent timeoffset=\\"10\\" message=\\"Welcome to: #8\\">
      </textevent>
      <textevent timeoffset=\\"360\\" message=\\"Let&apos;s talk about today&apos;s workout\\">
      </textevent>
      <textevent timeoffset=\\"370\\" message=\\"We&apos;ll be doing 5 blocks of VO2max work\\">
      </textevent>
      <textevent timeoffset=\\"380\\" message=\\"Each block 2 minutes and 30 seconds\\">
      </textevent>
      <textevent timeoffset=\\"390\\" message=\\"30 seconds more than last time :)\\">
      </textevent>
      <textevent timeoffset=\\"400\\" message=\\"Finally we finish off with 10 minutes of threshold work\\">
      </textevent>
      <textevent timeoffset=\\"410\\" message=\\"Easy as that :)\\">
      </textevent>
    </Warmup>
    <SteadyState Duration=\\"30\\" Power=\\"0.95\\" Cadence=\\"95\\">
      <textevent timeoffset=\\"0\\" message=\\"Spin to 95 RPM if you&apos;re not already there\\">
      </textevent>
    </SteadyState>
    <SteadyState Duration=\\"30\\" Power=\\"0.5\\" Cadence=\\"85\\">
      <textevent timeoffset=\\"0\\" message=\\"Rest a bit\\">
      </textevent>
    </SteadyState>
    <SteadyState Duration=\\"150\\" Power=\\"1.15\\" Cadence=\\"105\\">
      <textevent timeoffset=\\"0\\" message=\\"Here we go!\\">
      </textevent>
      <textevent timeoffset=\\"60\\" message=\\"First minute done\\">
      </textevent>
      <textevent timeoffset=\\"120\\" message=\\"Just 30 seconds more\\">
      </textevent>
      <textevent timeoffset=\\"140\\" message=\\"Finish it!\\">
      </textevent>
    </SteadyState>
    <SteadyState Duration=\\"180\\" Power=\\"0.5\\" Cadence=\\"85\\">
      <textevent timeoffset=\\"0\\" message=\\"Good!\\">
      </textevent>
      <textevent timeoffset=\\"10\\" message=\\"Rest for 3 minutes\\">
      </textevent>
      <textevent timeoffset=\\"170\\" message=\\"Ready for next round?\\">
      </textevent>
    </SteadyState>
    <Cooldown Duration=\\"450\\" PowerLow=\\"0.9\\" PowerHigh=\\"0.25\\">
      <textevent timeoffset=\\"0\\" message=\\"We&apos;ll now gradually decrease the power\\">
      </textevent>
      <textevent timeoffset=\\"10\\" message=\\"It&apos;s all downhill from here :)\\">
      </textevent>
      <textevent timeoffset=\\"20\\" message=\\"until you&apos;re cooled down\\">
      </textevent>
      <textevent timeoffset=\\"430\\" message=\\"Thanks for participating\\">
      </textevent>
      <textevent timeoffset=\\"440\\" message=\\"Until next time!\\">
      </textevent>
    </Cooldown>
  </workout>
</workout_file>"
`;

exports[`Generate ZWO examples/darth-vader.txt 1`] = `
"<workout_file>
  <name>Darth Vader</name>
  <author>HumanPowerPerformance.com</author>
  <description>Sign up for coaching with HumanPowerPerformance.com and get custom workouts and training plans</description>
  <sportType>bike</sportType>
  <workout>
    <SteadyState Duration=\\"300\\" Power=\\"0.55\\">
    </SteadyState>
    <IntervalsT Repeat=\\"10\\" OnDuration=\\"180\\" OnPower=\\"0.82\\" OffDuration=\\"20\\" OffPower=\\"1.3\\">
    </IntervalsT>
    <SteadyState Duration=\\"300\\" Power=\\"0.55\\">
    </SteadyState>
  </workout>
</workout_file>"
`;

exports[`Generate ZWO examples/ftp-test.txt 1`] = `
"<workout_file>
  <name>FTP Test (shorter)</name>
  <author>Zwift</author>
  <description>The short variation of the standard FTP test starts off with a short warmup,
a quick leg opening ramp, and a 5 minute hard effort to get the legs pumping.
After a brief rest it&apos;s time to give it your all and go as hard as you can for 20 solid minutes.
Pace yourself and try to go as hard as you can sustain for the entire 20 minutes -
you will be scored on the final 20 minute segment.

Upon saving your ride, you will be notified if your FTP improved.</description>
  <sportType>bike</sportType>
  <workout>
    <Warmup Duration=\\"300\\" PowerLow=\\"0.3\\" PowerHigh=\\"0.7\\">
      <textevent timeoffset=\\"20\\" message=\\"Welcome to the FTP test\\">
      </textevent>
      <textevent timeoffset=\\"35\\" message=\\"Time to get warmed up and get your cadence up to 90-100rpm\\">
      </textevent>
    </Warmup>
    <SteadyState Duration=\\"20\\" Power=\\"0.9\\">
    </SteadyState>
    <SteadyState Duration=\\"20\\" Power=\\"1.1\\">
    </SteadyState>
    <SteadyState Duration=\\"20\\" Power=\\"1.3\\">
    </SteadyState>
    <SteadyState Duration=\\"180\\" Power=\\"0.6\\">
    </SteadyState>
    <SteadyState Duration=\\"180\\" Power=\\"1.1\\">
    </SteadyState>
    <SteadyState Duration=\\"120\\" Power=\\"1.2\\">
    </SteadyState>
    <SteadyState Duration=\\"360\\" Power=\\"0.55\\">
      <textevent timeoffset=\\"10\\" message=\\"In 6 minutes the FTP test begins\\">
      </textevent>
      <textevent timeoffset=\\"300\\" message=\\"Prepare for your max 20 minute effort\\">
      </textevent>
    </SteadyState>
    <FreeRide Duration=\\"1200\\">
      <textevent timeoffset=\\"10\\" message=\\"Bring your power up to what you think you can hold for 20 minutes\\">
      </textevent>
      <textevent timeoffset=\\"300\\" message=\\"How ya doin?\\">
      </textevent>
      <textevent timeoffset=\\"600\\" message=\\"Half way done! If it&apos;s feeling easy now might be time to add 10 watts.\\">
      </textevent>
      <textevent timeoffset=\\"900\\" message=\\"5 minutes left!\\">
      </textevent>
      <textevent timeoffset=\\"960\\" message=\\"4 minutes left. Increase your power if you&apos;re feeling strong.\\">
      </textevent>
      <textevent timeoffset=\\"1020\\" message=\\"3 minutes left\\">
      </textevent>
      <textevent timeoffset=\\"1080\\" message=\\"2 minutes left\\">
      </textevent>
      <textevent timeoffset=\\"1140\\" message=\\"1 minute left. Go all in!\\">
      </textevent>
      <textevent timeoffset=\\"1170\\" message=\\"30 sec. Go for it.\\">
      </textevent>
      <textevent timeoffset=\\"1190\\" message=\\"10 seconds.\\">
      </textevent>
    </FreeRide>
    <Cooldown Duration=\\"300\\" PowerLow=\\"0.5\\" PowerHigh=\\"0.3\\">
    </Cooldown>
  </workout>
</workout_file>"
`;

exports[`Generate ZWO examples/halvfems.txt 1`] = `
"<workout_file>
  <name>Halvfems</name>
  <author></author>
  <description>Named after the number 90 in Danish, this workout is focused sweet spot training, centered around 90% of FTP.
This pairs with Devedeset (90 in Croatian) and Novanta (90 in Italian) to make up the sweet spot trifecta in this plan.
The workouts are alphabetically ordered from easiest to hardest, so enjoy the middle of these three challenging workouts today...</description>
  <sportType>bike</sportType>
  <workout>
    <Warmup Duration=\\"420\\" PowerLow=\\"0.25\\" PowerHigh=\\"0.75\\">
    </Warmup>
    <SteadyState Duration=\\"30\\" Power=\\"0.95\\" Cadence=\\"95\\">
    </SteadyState>
    <SteadyState Duration=\\"30\\" Power=\\"0.5\\" Cadence=\\"85\\">
    </SteadyState>
    <SteadyState Duration=\\"30\\" Power=\\"1.05\\" Cadence=\\"105\\">
    </SteadyState>
    <SteadyState Duration=\\"30\\" Power=\\"0.5\\" Cadence=\\"85\\">
    </SteadyState>
    <SteadyState Duration=\\"30\\" Power=\\"1.15\\" Cadence=\\"115\\">
    </SteadyState>
    <SteadyState Duration=\\"30\\" Power=\\"0.5\\" Cadence=\\"85\\">
    </SteadyState>
    <SteadyState Duration=\\"120\\" Power=\\"0.5\\" Cadence=\\"85\\">
    </SteadyState>
    <SteadyState Duration=\\"720\\" Power=\\"0.9\\" Cadence=\\"90\\">
    </SteadyState>
    <SteadyState Duration=\\"240\\" Power=\\"0.55\\" Cadence=\\"85\\">
    </SteadyState>
    <IntervalsT Repeat=\\"6\\" OnDuration=\\"60\\" OnPower=\\"0.9\\" Cadence=\\"60\\" OffDuration=\\"60\\" OffPower=\\"0.9\\" CadenceResting=\\"90\\">
      <textevent timeoffset=\\"0\\" message=\\"Start off at slow cadence\\">
      </textevent>
      <textevent timeoffset=\\"60\\" message=\\"Now bump that cadence up\\">
      </textevent>
      <textevent timeoffset=\\"120\\" message=\\"Back to 60 RPM\\">
      </textevent>
      <textevent timeoffset=\\"180\\" message=\\"Back to normal cadence\\">
      </textevent>
      <textevent timeoffset=\\"240\\" message=\\"Grinding...\\">
      </textevent>
      <textevent timeoffset=\\"300\\" message=\\"Back to normal cadence\\">
      </textevent>
      <textevent timeoffset=\\"360\\" message=\\"And grind at 60 RPM again\\">
      </textevent>
      <textevent timeoffset=\\"420\\" message=\\"Back to 90 RPM\\">
      </textevent>
      <textevent timeoffset=\\"480\\" message=\\"Lower it again to 60 RPM\\">
      </textevent>
      <textevent timeoffset=\\"540\\" message=\\"Increase the cadence to 90 RPM\\">
      </textevent>
      <textevent timeoffset=\\"600\\" message=\\"Last grinding minute\\">
      </textevent>
      <textevent timeoffset=\\"660\\" message=\\"Last minute... be strong!\\">
      </textevent>
    </IntervalsT>
    <SteadyState Duration=\\"240\\" Power=\\"0.55\\" Cadence=\\"85\\">
    </SteadyState>
    <IntervalsT Repeat=\\"4\\" OnDuration=\\"120\\" OnPower=\\"0.9\\" Cadence=\\"100\\" OffDuration=\\"60\\" OffPower=\\"0.9\\" CadenceResting=\\"65\\">
    </IntervalsT>
    <Cooldown Duration=\\"360\\" PowerLow=\\"0.55\\" PowerHigh=\\"0.25\\">
    </Cooldown>
  </workout>
</workout_file>"
`;

exports[`Generate ZWO examples/threshold-pushing.txt 1`] = `
"<workout_file>
  <name>Threshold pushing</name>
  <author>R.Saarsoo</author>
  <description>Start with a good warm up (10 minutes Zone 1, then 5 minutes Zone 2,
5 minutes Zone 3 followed by 5 minutes easy spinning).
Then straight in at Threshold Pushing powers for 12 mins.
Have 10 mins easy soft pedalling and repeat - warm down 15 minutes.
This session is designed to increase your FTP from below.
Working predominately aerobic metabolism.
In time your body will become more comfortable at these powers
and your FTP will increase.</description>
  <sportType>bike</sportType>
  <workout>
    <Warmup Duration=\\"1200\\" PowerLow=\\"0.45\\" PowerHigh=\\"0.75\\">
    </Warmup>
    <SteadyState Duration=\\"300\\" Power=\\"0.52\\">
    </SteadyState>
    <IntervalsT Repeat=\\"2\\" OnDuration=\\"720\\" OnPower=\\"1\\" Cadence=\\"90\\" OffDuration=\\"600\\" OffPower=\\"0.52\\">
    </IntervalsT>
    <Cooldown Duration=\\"600\\" PowerLow=\\"0.65\\" PowerHigh=\\"0.45\\">
    </Cooldown>
  </workout>
</workout_file>"
`;

exports[`Generate stats examples/comments.txt 1`] = `
"
Total duration: 24 minutes

Average intensity: 60%
Normalized intensity: 74%

TSS: 22

Zone Distribution:
 15 min - Z1: Recovery
  5 min - Z2: Endurance
  2 min - Z3: Tempo
  1 min - Z4: Threshold
  3 min - Z5: VO2 Max
  0 min - Z6: Anaerobic
"
`;

exports[`Generate stats examples/darth-vader.txt 1`] = `
"
Total duration: 43 minutes

Average intensity: 79%
Normalized intensity: 84%

TSS: 51

Zone Distribution:
 10 min - Z1: Recovery
  0 min - Z2: Endurance
 30 min - Z3: Tempo
  0 min - Z4: Threshold
  0 min - Z5: VO2 Max
  3 min - Z6: Anaerobic
"
`;

exports[`Generate stats examples/ftp-test.txt 1`] = `
"
Total duration: 45 minutes

Average intensity: 36%
Normalized intensity: 71%

TSS: 37

Zone Distribution:
 35 min - Z1: Recovery
  4 min - Z2: Endurance
  0 min - Z3: Tempo
  0 min - Z4: Threshold
  3 min - Z5: VO2 Max
  2 min - Z6: Anaerobic
"
`;

exports[`Generate stats examples/halvfems.txt 1`] = `
"
Total duration: 62 minutes

Average intensity: 74%
Normalized intensity: 81%

TSS: 68

Zone Distribution:
 22 min - Z1: Recovery
  2 min - Z2: Endurance
  0 min - Z3: Tempo
 37 min - Z4: Threshold
  1 min - Z5: VO2 Max
  0 min - Z6: Anaerobic
"
`;

exports[`Generate stats examples/threshold-pushing.txt 1`] = `
"
Total duration: 79 minutes

Average intensity: 69%
Normalized intensity: 78%

TSS: 81

Zone Distribution:
 42 min - Z1: Recovery
 13 min - Z2: Endurance
  0 min - Z3: Tempo
 24 min - Z4: Threshold
  0 min - Z5: VO2 Max
  0 min - Z6: Anaerobic
"
`;
